SpiceSharp/Algebra/Solve/ComplexSolver.cs

```diff
@@ -0,0 +1,333 @@
+                while (column.Row < step)
+                {
+                    element = Matrix.GetDiagonalElement(column.Row);
+
+                    // Mult = dest[row] / pivot
+                    mult = dest[column.Row].Value * element.Value;
+                    dest[column.Row].Value = mult;
+                    while ((element = element.Below) != null)
+                    {
+                        // dest[element.Row] -= mult * element
+                        dest[element.Row].Value -= mult * element.Value;
+                    }
+                    column = column.Below;
+                }
+
+                // Check for a singular matrix
+                element = Matrix.GetDiagonalElement(step);
+                if (element == null || element.Value.Equals(0.0))
+                    return false;
+                element.Value = Inverse(element.Value);
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Find the solution for a factored matrix and a right-hand-side
+        /// </summary>
+        /// <param name="solution">Solution vector</param>
+        public override void Solve(Vector<Complex> solution)
+        {
+            if (solution == null)
+                throw new ArgumentNullException(nameof(solution));
+
+            // TODO: Maybe we should cache intermediate
+            // Scramble
+            var intermediate = new Complex[Matrix.Size + 1];
+            var rhsElement = Rhs.First;
+            while (rhsElement != null)
+            {
+                intermediate[rhsElement.Index] = rhsElement.Value;
+                rhsElement = rhsElement.Next;
+            }
+
+            // Forward substitution
+            for (int i = 1; i <= Matrix.Size; i++)
+            {
+                Complex temp = intermediate[i];
+
+                // This step of the substitution is skipped if temp == 0.0
+                if (!temp.Equals(0.0))
+                {
+                    var pivot = Matrix.GetDiagonalElement(i);
+
+                    // temp = temp / pivot
+                    temp *= pivot.Value;
+                    intermediate[i] = temp;
+                    var element = pivot.Below;
+                    while (element != null)
+                    {
+                        // intermediate[row] -= temp * element
+                        intermediate[element.Row] -= temp * element.Value;
+                        element = element.Below;
+                    }
+                }
+            }
+
+            // Backward substitution
+            for (int i = Matrix.Size; i > 0; i--)
+            {
+                Complex temp = intermediate[i];
+                var pivot = Matrix.GetDiagonalElement(i);
+                var element = pivot.Right;
+
+                while (element != null)
+                {
+                    // temp -= element * intermediate[column]
+                    temp -= element.Value * intermediate[element.Column];
```

Comment:
@svenboulanger  Wouldn't be better to sum 'element.Value * intermediate[element.Column]' and then subtract only once? As far I as remember from university if temp is big number and 'element.Value * intermediate[element.Column]' is small there will be some truncating errors  when subtracting many times

Reply:
I don't really think it will make much difference because the same truncation errors can happen within the sum itself.